src/Course/StateT.hs

```diff
@@ -314,21 +275,22 @@ log1 =
 --
 -- >>> distinctG $ listh [1,2,3,2,6,106]
 -- Logger ["even number: 2","even number: 2","even number: 6","aborting > 100: 106"] Empty
-distinctG ::
-  (Integral a, Show a) =>
-  List a
-  -> Logger Chars (Optional (List a))
-distinctG =
-  error "todo: Course.StateT#distinctG"
-
-onFull ::
-  Applicative f =>
-  (t -> f (Optional a))
-  -> Optional t
-  -> f (Optional a)
+
+-- data Logger l a = Logger (List l) a deriving (Eq, Show)
+-- data OptionalT f a = OptionalT {runOptionalT :: f (Optional a)}
+
+distinctG :: (Integral a, Show a) => List a -> Logger Chars (Optional (List a))
+distinctG xs = filterEmptyness $ runOptionalT (evalT (filtering p (reverse xs)) S.empty)
+  where
+    filterEmptyness (Logger log vals) = Logger (filter (not . isEmpty) log) (reverse <$> vals)
+    logFilter x = if even x then "even number: " ++ show' x else Nil
+    p x
+      | x < 100 = StateT (\s -> OptionalT (log1 (logFilter x) (Full (S.notMember x s, S.insert x s))))
+      | otherwise = StateT (\s -> OptionalT (log1 ("aborting > 100: " ++ show' x) Empty))
```

Comment:
oh noes... `== 100` would abort...  :)

Reply:
oops, forgot that one, I want to refactor this a bit though, this was my first solution